Philip/daily bonus

mango-habanero opened the pull request:
This PR attempts to create a unique celery worker ( ```ge_worker``` ) that is entirely independent of the `eth_worker` responsible for blockchain tasks.
It includes a bonus processor that computes the daily bonus amount for users based on their percentage unique outward transactions in a 24hr span of time relative to the collective system's unique outward transactions.

mango-habanero committed:
Creates instance of celery.
- Adds a new instance of celery with a beat scheduler set to run every day at 11:50PM EAT to disburse daily bonus amounts.

mango-habanero committed:
Adds bonus processor
- Adds functionality to compute daily bonus for each user.

mango-habanero committed:
Defines celery task to issue daily bonus.

mango-habanero committed:
Removes blockchain address from query.
Make payment method requires user object and does not take a blockchain address.

mango-habanero committed:
Rudimentary set up for docker container.
- Dockerfile defines process to setup a worker [Mirrors eth_worker setup and may require refactoring]
- start_beat_scheduler starts a cron job for the issuance of daily bonuses.
- start_bonuses_worker starts worker.

mango-habanero commented on ge_worker/Dockerfile:
```diff
@@ -0,0 +1,18 @@
+
+RUN chmod +x /_start_bonuses_worker.sh
+RUN chmod +x /_start_beat_scheduler.sh
+
```
@nolash I may need some help with setting up this docker container needed to run the issuing of the daily bonus, this one mirrors the eth_worker dockerfile but is untested.

nolash replied:
It would be good to have a session with @kibyegn on deploying this. That way we can make sure that we have the same understanding among the three of us on what procedure and conventions we use (I'm not an advanced user of docker). Before that I guess that this is exactly the kind of thing we need bootstrapped local dev env for, so that eg I can do a certain degree of verification locally on my own computer while reviewing the solution.

mango-habanero committed:
Refactors function performing db query.

mango-habanero committed:
Refactors import.

mango-habanero committed:
Refactors celery app setup

mango-habanero committed:
Refactors start of bonuses worker.

mango-habanero commented on ge_worker/bonuses/bonus_processor.py:
```diff
@@ -68,28 +93,49 @@ class BonusProcessor:
+                                      receive_user=user,
+                                      transfer_subtype=TransferSubTypeEnum.DISBURSEMENT)
+
+                db.session.commit()
```
@nolash Still wrapping my head around the whole celery dynamic but these changes slightly optimize the daily bonus calculation.
Unfortunately the ```make_payment_transfer``` functions seems to spawn a synchronous task which raises and issues. So that's the next thing I need to figure out

nolash replied:
@PhilipWafula is `make_payment_transfer` a blocker then? Can you please make an issue for this, describing the exact nature of the problem?

mango-habanero replied:
Using the format you proposed I imagine ?

nolash commented on ge_worker/_start_bonuses_worker.sh:
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+
+celery -A bonuses worker --loglevel=INFO
```
Do we need both the `beat` one and this one?

mango-habanero replied:
In terms of functionality yes one starts the worker, the other starts  the beat scheduler, but in terms of scripts , I could run them from them same script, again this was a choice in view of being consistent with the Sempo build.

nolash replied:
Sorry the question came from ignorance, I read up a bit and apparently both are needed for this.

nolash commented on ge_worker/bonuses/bonus_tasks.py:
```diff
@@ -0,0 +1,8 @@
+
+@celery_app.task()
+def auto_disburse_daily_bonus():
+    bonus_processor = BonusProcessor(issuable_amount=1000)
```
Should this be hardcoded to `1000`?

mango-habanero replied:
The idea is to have it configurable but we need to agree on whether we're setting a GUI for the rewards system before that can be baked into this. It should come as part of this [Write spec for rewards system #13](https://github.com/GrassrootsEconomics/CIC-Docs/issues/13)
Consider it a test value until we're agreed on how to proceed on that.
Essentially the final `Bonus processor` should allow for configurable time frames as well as minimum amounts.

nolash commented on ge_worker/bonuses/bonus_processor.py:
```diff
@@ -0,0 +1,141 @@
+from flask import g
+from logging import Logger
+
+parent_dir = os.path.abspath(os.path.join(os.getcwd(), "../app"))
```
`os.getcwd()`

is this dangerous? Should we get the script path instead?

mango-habanero replied:
This implementation is done with conventions of the Sempo build in mind. I was experimenting with a different flask project [see setup script](https://github.com/PhilipWafula/faulu-apis-demo/blob/master/setup.sh) and I decided to use the approach we had discussed while in Kilifi.
I'd argue that adding the root folder of the whole application to the python path would help us avoid the numerous occurrences of this in the code. It is really irksome having to add this every time I need to access that server package.
@nolash what do you think ?

nolash replied:
>  It is really irksome having to add this every time I need to access that server package.

Yeah it seems like bad design. I concur.

nolash commented on ge_worker/bonuses/bonus_processor.py:
```diff
@@ -0,0 +1,141 @@
+        - count unique recipient user ids.
+        - credit transfer amount is greater or equal to 20wei.
+        - credit transfer occurs since 24hrs ago.
+        - credit transfer status is COMPLETE
```
Does this mean that any `PENDING` transactions within the time frame will be missed?

mango-habanero replied:
my argument would be that a pending transaction within the time frame does not add to the transaction volume until it is complete, if it completes after the the time frame then it becomes a complete transaction for the next cycle.
@WillRuddick thoughts ?

nolash replied:
In that case we will need to scope the time period to `updated` instead of `created` (and make sure that `updated` actually is sane)

nolash commented on ge_worker/bonuses/bonus_processor.py:
```diff
@@ -0,0 +1,141 @@
+    FROM credit_transfer
+    LEFT JOIN transfer_account 
+    ON credit_transfer.sender_transfer_account_id = transfer_account .id
+    WHERE credit_transfer._transfer_amount_wei >= (2*10e17)
```
Please consider replacing `2*10e7` literal with a config parameter.

mango-habanero replied:
Will these config parameters be editable outside of the code ? Because the time and amount values here are eventually intended to be configurable using a GUI or with arguments provided in the cli depending on what approach we settle on.

nolash replied:
I think they should at a minimum be in platform config.

nolash commented on ge_worker/bonuses/bonus_processor.py:
```diff
@@ -0,0 +1,141 @@
+    AND credit_transfer.transfer_status = 'COMPLETE'
+    AND credit_transfer.transfer_subtype = 'STANDARD'
+    GROUP BY 
+    credit_transfer.sender_user_id'''.format((datetime.now() - timedelta(hours=24)))
```
Please consider replacing `timedelta.hours` literal with a config parameter. Also please consider making the calculation a local variable above. This is a matter of taste, I know, but let me suggest a convention that for readability reasons we keep inlined calculations, chainings and casts to a minimum.

nolash commented on ge_worker/bonuses/bonus_processor.py:
```diff
@@ -0,0 +1,141 @@
+    credit_transfer.sender_user_id,
+    COUNT (DISTINCT (credit_transfer.recipient_user_id))
+    FROM credit_transfer
+    LEFT JOIN transfer_account 
```
Why `LEFT JOIN`  and not `INNER JOIN`?

mango-habanero replied:
This I should fix.

nolash commented on ge_worker/bonuses/bonus_processor.py:
```diff
@@ -0,0 +1,141 @@
+    AND credit_transfer.created > '{}'
+    AND credit_transfer.transfer_status = 'COMPLETE'
+    AND credit_transfer.transfer_subtype = 'STANDARD'
+    GROUP BY 
```
Please consider whether there is any merit to splitting this into two different job cycles. First the count of recipients, then individual queries for each transfer amount calculation. I am not familiar with postgresql db internals, but I assume that multiple queries with low priority would have less an impact on db availablity for time-criticial tasks than one big task with low priority.

Such a solution would mean async individual queries and a listener to determine when all the calculations have been done.

Alternately, let's verify that there is a way to perform postgresql queries which can yield mid-process to higher  priority tasks (this might even be a good stackoverflow question).

mango-habanero replied:
This may require some research but it does pose a really good question, I will look into it.

nolash commented on ge_worker/bonuses/bonus_processor.py:
```diff
@@ -0,0 +1,141 @@
+            user_bonus_amount = int(user_percentage_unique_outward_transactions * self.issuable_amount)
+
+            # define logger
+            print('USER: {} {}, '
```
Please let's use `logging`

nolash reviewed (changes requested):
How about unit test code?

mango-habanero committed:
Refactors bonus processor
- Adds config params to get_list_of_users_with_total_unique_outward_transactions
- Move calculation to local variables
- Adds celery task logging
- Replaces LEFT JOIN with INNER JOIN

mango-habanero committed:
Merge branch 'master' of github.com:teamsempo/SempoBlockchain into philip/daily_bonus

mango-habanero committed:
Adds scripts for developer sensitive operations.

mango-habanero committed:
Removes ge_worker

mango-habanero committed:
Refactors script to start beat scheduler

mango-habanero committed:
Adds rudimentary tests for rewards processor

mango-habanero committed:
Adds rewards system configs

mango-habanero committed:
Adds auto disbursement task to eth_worker

mango-habanero committed:
Adds rewards system

mango-habanero committed:
Adds script to seed transactions.

mango-habanero committed:
Updates npm packages

enjeyw committed:
Merge branch 'master' into philip/daily_bonus

enjeyw commented on app/server/utils/rewards/queries.py:
```diff
@@ -0,0 +1,52 @@
+    credit_transfer.sender_user_id'''.format(transfer_amount, created_since)
+
+    # execute query
+    result = db.session.execute(sql_query)
```
Might be good to use Sqlalchemy's text function to avoid any chance of injection:
https://docs.sqlalchemy.org/en/13/core/tutorial.html#using-textual-sql

nolash commented on app/server/api/rewards_api.py:
```diff
@@ -0,0 +1,50 @@
+rewards_view = DisburseDailyBonusAPI.as_view('rewards_api_view')
+rewards_blueprint.add_url_rule('/rewards/disburse_daily_bonuses/',
+                               view_func=rewards_view,
+                               methods=['POST', ])
```
dont need the last comma

nolash commented on app/server/utils/rewards/queries.py:
```diff
@@ -0,0 +1,52 @@
+from server import db
+
+
+def users_with_corresponding_total_outward_txs(created_since: int,
```
Would it not be better to integrate with `app/server/model/transfer_account.py` here instead? it already has a list of credit transfer as a relation.

mango-habanero replied:
I tried running a scenario with this, I don't think there's a change in overhead @nolash but I am curios what did you have in mind with this suggestion ?

nolash commented on config.py:
```diff
@@ -352,3 +352,7 @@ def address_from_private_key(private_key):
     GE_DB_PASSWORD = ''
     GE_HTTP_PROVIDER = ''
+
+TRANSFERS_CREATED_SINCE = common_secrets_parser['REWARDS'].getint('created_since')
```
Why secret?

Is `created_since` a quasi constant value?

mango-habanero replied:
In the meantime yes. Plus I imagine data on our rewards systems is sensitive information

nolash commented on eth_worker/eth_manager/__init__.py:
```diff
@@ -61,6 +62,13 @@
+celery_app.conf.beat_schedule = {
+    "manage_daily_bonuses": {
+        "task": "eth_manager.rewards_celery_tasks.disburse_daily_bonuses",
+        "schedule": crontab(minute="50", hour="23")
```
hour/minute in config file?

nolash commented on test/unit/app/utils/rewards/test_processor.py:
```diff
@@ -0,0 +1,23 @@
+from app.server.utils.rewards.processor import get_admin_account_to_disburse
+
+
+def test_get_user_by_id(test_client, init_database, create_transfer_account_user):
```
Here you are really only testing the sqlalchemy query filter. I think it is not necessary.

nolash commented on test/unit/app/utils/rewards/test_processor.py:
```diff
@@ -0,0 +1,23 @@
+from app.server.utils.rewards.processor import get_user_by_id
```
I can't really see any test that checks whether the disbursement amounts are correct? Are those tests somewhere else?

nolash commented on eth_worker/eth_manager/rewards_celery_tasks.py:
```diff
@@ -0,0 +1,26 @@
+        'transfer_amount': transfer_amount
+    }
+
+    response = requests.post(app_host + '/api/v1/rewards/disburse_daily_bonuses/',
```
I'm not sure I understand. Is the celery task only used as a cron job here? It doesn't seem to actually process tasks in a queue?

nolash commented on app/server/models/reward.py:
```diff
@@ -0,0 +1,24 @@
+class Reward(ModelBase):
+    __tablename__ = 'reward'
+
+    tag = db.Column(db.String)
```
A comment explaining what `tag` and `recipients_data` is here would be nice.

nolash commented on app/server/utils/rewards/processor.py:
```diff
@@ -0,0 +1,182 @@
+
+def get_admin_account_to_disburse(subject_user: User):
+    """
+    This method gets all admin users for a specific organization and uses the first admin account to disburse a specific
```
How do you know using the first one is correct?

nolash commented on app/server/models/reward.py:
```diff
@@ -0,0 +1,24 @@
+    __tablename__ = 'reward'
+
+    tag = db.Column(db.String)
+    recipients_data = db.Column(JSONB, default={})
```
I would prefer we not use (more) json types in the db please.

mango-habanero replied:
@nolash Why not ? 
Also in this particular instance, I think I JSON is a good idea because different rewards may collect different user data. For instance, daily bonuses collect the following:
``` 
'user_ids_with_total_unique_outward_txs' : 'some_value',
`collective_unique_outward_txs`: 'some_value'
```

Maybe a reward involving chamas will collect a different format of data for instance: 
```
'chama_tx_volume': 'some_value',
'group_membership_ratio': 'some_value',
'user_ids_with_total_unique_outward_txs': 'some_value'
```

nolash commented on app/server/utils/rewards/processor.py:
```diff
@@ -0,0 +1,182 @@
+    for counter, \
+        (user,
+         unique_outward_transactions_per_user,
+         disbursing_admin,
```
Won't `disbursing_admin` be the same here always

PhilipWafula replied:
It actually should come from the organization level account so I need to correct this

nolash commented on app/server/utils/rewards/processor.py:
```diff
@@ -0,0 +1,182 @@
+        daily_bonus_query_result=daily_bonus_query_result)
+
+    # iterate through list to make disbursements
+    for counter, \
```
If this loop dies, how do we know where we left off to start again?

nolash commented on app/server/utils/rewards/processor.py:
```diff
@@ -0,0 +1,182 @@
+    db.session.commit()
+
+
+def process_daily_bonus_query(daily_bonus_query_result: list):
```
Perhaps this should be a temporary table?

mango-habanero replied:
I don't think so, I'd argue that a temporary table would be overkill for a one time process that isn't as resource heavy. But if you could please elaborate on this suggestion.

nolash commented on app/server/utils/rewards/processor.py:
```diff
@@ -0,0 +1,182 @@
+        user_bonus_amount = int(user_percentage_unique_outward_transactions * issuable_amount)
+
+        if user_bonus_amount > 1:
+            disbursement = make_payment_transfer(user_bonus_amount,
```
Is this an asyncronous call? If not, we could potentially wait a loooong time on this.

mango-habanero committed:
Adds beat scheduler crontab values to config

mango-habanero committed:
Removes auxiliary test

mango-habanero committed:
Refactors processor

- Changes method to get token from indexed column using token id retrieved from query.
- Gets single disbursing admin to use for all disbursements.
- Adds list of supported REWARDS_DATA_TAGS
- Adds logging for transactions.
-Adds persistence for disbursement failure points to provide a return checkpoint when issuing daily bonuses.

mango-habanero committed:
Refactors query

- Adds token ifs from transfer accounts.
- Wraps query in text to prevent injection attacks.

mango-habanero committed:
Minor cleanup

mango-habanero committed:
Merge branch 'philip/daily_bonus' of github.com:GrassrootsEconomics/SempoBlockchain into philip/daily_bonus

mango-habanero committed:
Merge branch 'master' into philip/daily_bonus

mango-habanero committed:
Moves reward configs to public configs

mango-habanero committed:
Adds reward configs  to docker test config file.

nolash commented:
Work will be re-submitted in gitlab
